pkg/util/winutil/scmmonitor.go

```diff
@@ -0,0 +1,203 @@
+
+	//hdr.Data = unsafe.Pointer(&buf[0])
+	//hdr.Len = int(servicesReturned)
+	//hdr.Cap = int(servicesReturned)
+
+	newmap := make(map[uint64]*Service)
+	for idx, svc := range services {
+		var thissvc *Service
+		var ok bool
+		if thissvc, ok = newmap[uint64(svc.ServiceStatusProcess.ProcessId)]; !ok {
+			thissvc = &Service{}
+			newmap[uint64(svc.ServiceStatusProcess.ProcessId)] = thissvc
+		}
+		thissvc.essp = append(thissvc.essp, &services[idx])
+	}
+	var current windows.Filetime
+	windows.GetSystemTimeAsFileTime(&current)
+	scm.lastMapTime = uint64(current.Nanoseconds())
+	scm.pidToService = newmap
+	scm.serviceRefreshes++
+	return nil
+}
+
+func (s *Service) toServiceInfo() *ServiceInfo {
+	var si ServiceInfo
+	for _, inf := range s.essp {
+		si.DisplayName = append(si.DisplayName, windows.UTF16PtrToString(inf.DisplayName))
+		si.ServiceName = append(si.ServiceName, windows.UTF16PtrToString(inf.ServiceName))
+	}
+	return &si
+
+}
+
+func (scm *SCMMonitor) getServiceFromCache(pid, pidstart uint64) (*ServiceInfo, error) {
+	var err error
+	if val, ok := scm.pidToService[pid]; ok {
+		// it was in there.  see if it's the right one
+		if val.startTime == 0 {
+			// we don't prepopulate the cache with pid start times
+			val.startTime, err = pGetProcessStartTimeAsNs(pid)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if val.startTime == pidstart {
+			// it's the same one.
+			return val.toServiceInfo(), nil
+		}
+	}
+	return nil, nil
+}
+
+// GetServiceInfo gets the service name and display name if the process identified
+// by the pid is in the SCM.  A process which is not an SCM controlled service will
+// return nil with no error
+func (scm *SCMMonitor) GetServiceInfo(pid uint64) (*ServiceInfo, error) {
+	// get the process start time of the pid being checked
+	pidstart, err := pGetProcessStartTimeAsNs(pid)
+	if err != nil {
+		return nil, err
+	}
+	scm.mux.Lock()
+	defer scm.mux.Unlock()
+	// check to see if the pid is in the cache of known, not service pids
+	if val, ok := scm.nonServicePid[pid]; ok {
+		// it's a known non service pid.  Make sure it's not recycled.
+		if val == pidstart {
+			// it's the same process.  We know this isn't a service
+			return nil, nil
+		} else {
+			// it was in there but the times didn't match, which means
+			// it's a different process.  Clean it out.
+			delete(scm.nonServicePid, pid)
+		}
+	}
+	// if we get here it either wasn't in the map, or it was but the
+	// start time didn't match
+
+	if pidstart <= scm.lastMapTime {
+		// so it's been around longer than our last check of the service
+		// table.  so if it's in there it's probably good.
+		si, err := scm.getServiceFromCache(pid, pidstart)
+		if (si == nil && err != nil) || (si != nil && err == nil) {
+			return si, err
+		}
+		// else
+		scm.nonServicePid[pid] = pidstart
+		return nil, nil
+
+	}
+
+	// if we get here, either it wasn't in either map, or the process
```

Comment:
I feel that the comment is misleading at least to the code above. It looks like we get here only if process is newer than last cache time